src/common/ceph_mutex.h

```diff
@@ -107,6 +107,47 @@ namespace ceph {
   #define ceph_mutex_is_not_locked_by_me(m) (!(m).is_locked_by_me())
 }
 
+#elif __MINGW32__
+
+// The winpthreads shared mutex implementation is broken.
+// We'll use boost::shared_mutex instead.
+// https://github.com/msys2/MINGW-packages/issues/3319
+
+#include <boost/thread/shared_mutex.hpp>
```

Comment:
Technically, `<condition_variable>` and `<mutex>` need to be included here.  And that would make it really clear that it's a cut-and-paste from the "release (fast and minimal)" branch.  How about something like:
```
diff --git a/src/common/ceph_mutex.h b/src/common/ceph_mutex.h
index 8afdf922eb76..028aeae0f00c 100644
--- a/src/common/ceph_mutex.h
+++ b/src/common/ceph_mutex.h
@@ -156,27 +156,39 @@ namespace ceph {
 
 #include <condition_variable>
 #include <mutex>
-#include <shared_mutex>
 
+// The winpthreads shared mutex implementation is broken.
+// We'll use boost::shared_mutex instead.
+// https://github.com/msys2/MINGW-packages/issues/3319
+#if __MINGW32__
+#include <boost/thread/shared_mutex.hpp>
+#else
+#include <shared_mutex>
+#endif
 
 namespace ceph {
 
   typedef std::mutex mutex;
   typedef std::recursive_mutex recursive_mutex;
   typedef std::condition_variable condition_variable;
+
+#if __MINGW32__
   typedef std::shared_mutex shared_mutex;
+#else
+  typedef boost::shared_mutex shared_mutex;
+#endif
 
   // discard arguments to make_mutex (they are for debugging only)
   template <typename ...Args>
-  std::mutex make_mutex(Args&& ...args) {
+  mutex make_mutex(Args&& ...args) {
     return {};
   }
   template <typename ...Args>
-  std::recursive_mutex make_recursive_mutex(Args&& ...args) {
+  recursive_mutex make_recursive_mutex(Args&& ...args) {
     return {};
   }
   template <typename ...Args>
-  std::shared_mutex make_shared_mutex(Args&& ...args) {
+  shared_mutex make_shared_mutex(Args&& ...args) {
     return {};
   }

```

Reply:
Makes sense, thanks for the suggestion. I’m currently on vacation but I’ll update the PR next week.

Reply:
i second Ilya's comment.